File: robocode-tankroyale-server/src/main/java/dev/robocode/tankroyale/server/util/ColorUtil.java
package dev.robocode.tankroyale.server.util;

public final class ColorUtil {

  public static Integer colorStringToRGB(String colorStr) {
    if (colorStr == null) {
      return null;
    }
    colorStr = colorStr.trim();
    if (!RegExp.HEX_DEC_COLOR_CODE.matcher(colorStr).matches()) {
      return null;
    }
    if (colorStr.length() == 7) {
      return hexToRgb16bit(colorStr);
    }
    if (colorStr.length() == 4) {
      return hexToRgb8bit(colorStr);
    }
    return null;
  }

  public static int hexToRgb16bit(String colorStr) {
    int r = Integer.valueOf(colorStr.substring(1, 3), 16);
    int g = Integer.valueOf(colorStr.substring(3, 5), 16);
    int b = Integer.valueOf(colorStr.substring(5, 7), 16);

    return r << 16 | g << 8 | b;
  }

  public static int hexToRgb8bit(String colorStr) {
    int r = Integer.valueOf(colorStr.substring(1, 2), 16);
    int g = Integer.valueOf(colorStr.substring(2, 3), 16);
    int b = Integer.valueOf(colorStr.substring(3, 4), 16);

    r = r << 4 | r;
    g = g << 4 | g;
    b = b << 4 | b;

    return r << 16 | g << 8 | b;
  }
}

File: gui-app/src/main/kotlin/dev/robocode/tankroyale/gui/util/ColorUtil.kt
package dev.robocode.tankroyale.gui.util

import java.awt.Color
import java.util.regex.Pattern

class ColorUtil {
    companion object {
        private val NUMERIC_RGB = Pattern.compile("^#[0-9a-fA-F]{3,6}$")

        fun fromString(str: String): Color {
            require(NUMERIC_RGB.matcher(str).matches()) { "Illegal color format: $str" }
            val r = Integer.valueOf(str.substring(1, 3), 16)
            val g = Integer.valueOf(str.substring(3, 5), 16)
            val b = Integer.valueOf(str.substring(5, 7), 16)
            return Color(r, g, b)
        }
    }
}